File: uk.ac.gda.server.exafs/scripts/exafsscripts/exafs/b18OutputPreparer.py
from metadata import Metadata
from gdascripts.metadata.metadata_commands import meta_clear
    
class B18OutputPreparer:
    
    def __init__(self, datawriterconfig):
        self.datawriterconfig = datawriterconfig
    
    def prepare(self, outputParameters, scanBean):
        metadata = outputParameters.getMetadataList()
        meta=Metadata(self.datawriterconfig)
        if len(metadata)>0:
            meta.add_to_metadata(metadata)
        
    # Determines the AsciiDataWriterConfiguration to use to format the header/footer of the ascii data files
    # If this returns None, then let the Ascii Data Writer class find the config for itself.
    def getAsciiDataWriterConfig(self,scanBean):
        return self.datawriterconfig

    # For any specific plotting requirements based on all the options in this experiment
    def getPlotSettings(self,detectorBean,outputBean):
        return None
    
    def _resetHeader(self):
        self.datawriterconfig.setHeader(self.original_header)
        meta_clear()

File: uk.ac.gda.server.exafs/src/uk/ac/gda/server/exafs/scan/preparers/B18OutputPreparer.java
package uk.ac.gda.server.exafs.scan.preparers;

import gda.data.metadata.NXMetaDataProvider;
import gda.data.scan.datawriter.AsciiDataWriterConfiguration;
import gda.device.DeviceException;
import gda.factory.Finder;
import gda.scan.ScanPlotSettings;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import uk.ac.gda.beans.exafs.IDetectorParameters;
import uk.ac.gda.beans.exafs.IOutputParameters;
import uk.ac.gda.beans.exafs.IScanParameters;
import uk.ac.gda.beans.exafs.MetadataParameters;
import uk.ac.gda.server.exafs.scan.OutputPreparer;

public class B18OutputPreparer implements OutputPreparer {

	private AsciiDataWriterConfiguration datawriterconfig;
	private NXMetaDataProvider metashop;
	private Set<String> scannablesAddedToMetadata = new HashSet<String>();
	
	public B18OutputPreparer(AsciiDataWriterConfiguration datawriterconfig, NXMetaDataProvider metashop) {
		this.datawriterconfig = datawriterconfig;
		this.metashop =  metashop;
	}

	@Override
	public void prepare(IOutputParameters outputParameters, IScanParameters scanBean) throws DeviceException {
		List<MetadataParameters> metadata = outputParameters.getMetadataList();
		for(MetadataParameters parameter : metadata) {
			if (!metashop.containsKey(parameter.getScannableName())) {
				metashop.add(Finder.getInstance().find(parameter.getScannableName()));
				scannablesAddedToMetadata.add(parameter.getScannableName());
			}
		}
	}

	// # Determines the AsciiDataWriterConfiguration to use to format the header/footer of the ascii data files
	// # If this returns None, then let the Ascii Data Writer class find the config for itself.
	@Override
	public AsciiDataWriterConfiguration getAsciiDataWriterConfig(IScanParameters scanBean) {
		return datawriterconfig;
	}

	@Override
	public void _resetNexusStaticMetadataList() {
		// do nothing
		for (String scannable : scannablesAddedToMetadata) {
			metashop.remove(scannable);
		}
		scannablesAddedToMetadata.clear();
	}

	@Override
	public ScanPlotSettings getPlotSettings(IDetectorParameters detectorBean, IOutputParameters outputBean) {
		return null;
	}

}